compiler/src/lnntoamm/Types.ts

```diff
@@ -85,54 +103,85 @@ export class Builtin extends Type {
     return this;
   }
 
-  compatibleWithConstraint(ty: Type): boolean {
+  compatibleWithConstraint(ty: Type, scope: Scope): boolean {
     if (ty instanceof Builtin) {
       return this.eq(ty);
-    } if (ty instanceof OneOf || ty instanceof Generated) {
-      return ty.compatibleWithConstraint(this);
+    } else if (ty instanceof OneOf || ty instanceof Generated) {
+      return ty.compatibleWithConstraint(this, scope);
+    } else if (ty instanceof HasOperator) {
+      return Has.operator(ty, scope, this).length !== 0;
+    } else if (ty instanceof HasMethod) {
+      return Has.method(ty, scope, this).length !== 0;
+    } else if (ty instanceof HasField) {
+      return false;
```

Comment:
Should there be a `//TODO: Once structs are implemented fix this` here?

Reply:
er, not here - builtins never have user-accessible fields right?

Reply:
Unfortunately no. `ExecRes`, `InitialReduce`, `KeyVal`, `InternalRequest`, and `InternalResponse` all have user-accessible fields but are also built-in types as some opcodes directly consume or produce them.

Reply:
gotcha, i'll leave a TODO then